---
title: "Dataset Cleanup Tutorial"
output:
  html_document: default
---

## Intro

**This script is a guide for uniformly formatting datasets for Weighted Gene Co-expression Network Analysis (WGCNA)**
The examples in this tutorial will be from GEO datasets GSE156037, GSE41205, GSE63528, and GSE54853. 
Each dataset presents a slightly different cleanup challenge.
- GSE156037 -- convert from gene stable IDs to transcript stable IDs
- GSE41205 -- merge 2 samples with different number of genes, select only the counts columns
- GSE63528 -- merge 5 samples with same number of genes, select only the counts columns
- GSE46257 -- convert gene names to transcript stable IDs, select only counts columns

### Setup

Set working directory

**Change directory paths as needed throughout script**
```{r set directory}
knitr::opts_knit$set(root.dir = '~/bioinformatics/group-projects-invivo-fall-project/dataset_cleanup_tutorial/')
```

Load libraries
```{r libraries, message=FALSE}
library(tidyverse)
library(readxl)
```

### Read in Datasets
1 dataset is a csv file, 2 are tsv files, and 1 is an xls file. 

Aptly, the `read.csv("file")` function reads in csv files, and `read_xls("file")` reads in xls files.

To read in tsv files, use `read.delim("file", sep="\t")`. Note, the functions reading in GSE63528 samples include the parameter `row.names = 1`, while those for reading in GSE41205 samples do not. Including `row.names = 1` tells R to take column 1 and specify it as the rownames of the dataframe. This allows us to retain the gene information when selecting and merging only the counts column simultaneously from the 5 sample datasets. This will make more sense in a few steps.

For each dataset check what kind of ID there is, if there are other columns besides raw gene counts, and the initial number of genes. Many of the GEO datasets selected have Transcript Stable IDs (F07C3.7) but some have Gene Stable IDs (WBGene00000003) or Gene Names (aat-2). Converting to Transcript Stable IDs conserved the most information so that is what this script will do.

```{r GSE156037 csv, results='hold'}
GSE.156037 <- read.csv("../Selected_GEO_datasets_raw/GSE156037_raw_counts.csv")

# check what kind of IDs, if there are other columns besides raw counts, and number of genes
print(head(GSE.156037))
sprintf("initial number of genes: %i", nrow(GSE.156037))
```

```{r GSE41205 tsv, results='hold'}
GSE.41205_N2 <- read.delim("../Selected_GEO_datasets_raw/GSE41205_Paul_2012_MAPK/GSM1010460_N2genes.txt", 
                                sep = "\t")
GSE.41205_KU25 <- read.delim("../Selected_GEO_datasets_raw/GSE41205_Paul_2012_MAPK/GSM1010461_KU25genes.txt", 
                                  sep = "\t")

print(head(GSE.41205_N2))
print(head(GSE.41205_KU25))
sprintf("initial number of genes: %i", nrow(GSE.41205_N2))
sprintf("initial number of genes: %i", nrow(GSE.41205_KU25))
```

```{r GSE63528 tsv, results='hold'}
GSE.63528_1 <- read.delim("../Selected_GEO_datasets_raw/GSE63528_Rangaraju_2016/GSM1551786_Sample_1_processed_data.txt", 
                          sep = "\t", row.names = 1)
GSE.63528_2 <- read.delim("../Selected_GEO_datasets_raw/GSE63528_Rangaraju_2016/GSM1551787_Sample_2_processed_data.txt", 
                          sep = "\t", row.names = 1)
GSE.63528_3 <- read.delim("../Selected_GEO_datasets_raw/GSE63528_Rangaraju_2016/GSM1551788_Sample_3_processed_data.txt", 
                          sep = "\t", row.names = 1)
GSE.63528_4 <- read.delim("../Selected_GEO_datasets_raw/GSE63528_Rangaraju_2016/GSM1551789_Sample_4_processed_data.txt", 
                          sep = "\t", row.names = 1)
GSE.63528_5 <- read.delim("../Selected_GEO_datasets_raw/GSE63528_Rangaraju_2016/GSM1551790_Sample_5_processed_data.txt", 
                          sep = "\t", row.names = 1)

print(head(GSE.63528_1))
print(head(GSE.63528_2))
print(head(GSE.63528_3))
print(head(GSE.63528_4))
print(head(GSE.63528_5))
sprintf("initial number of genes: %i", nrow(GSE.63528_1))
sprintf("initial number of genes: %i", nrow(GSE.63528_2))
sprintf("initial number of genes: %i", nrow(GSE.63528_3))
sprintf("initial number of genes: %i", nrow(GSE.63528_4))
sprintf("initial number of genes: %i", nrow(GSE.63528_5))
```

```{r GSE46257 xls, results='hold'}
GSE.46257 <- read_xls("../Selected_GEO_datasets_raw/GSE46257_Schmeisser_2013_arsenite/GSE46257_ce_Nr35-38_counts_rpkm.xls")

print(head(GSE.46257))
sprintf("initial number of genes: %i", nrow(GSE.46257))
```


### Cleanup

Read in ID table containing Gene Stable IDs, Transcript Stable IDs, and Gene Name
```{r}
ID_table <- read.delim("../Cleaned_Datasets_transcript_stableID/celegans_cleaned_mart_ensembl.102.tsv", sep = "\t")
head(ID_table)
```

Convert gene stable IDs to transcript stable IDs
```{r GSE156037 cleanup}
head(GSE.156037)  # check the name of the ID column
                                                  # change by.x to match gene ID column
GSE.156037_converted <- merge(GSE.156037, ID_table, by.x = "rowname", by.y = "Gene.stable.ID", all.x = FALSE, all.y = FALSE)
head(GSE.156037_converted)  # check which columns you want to keep, i.e. only transcript stable ID and raw counts columns

GSE.156037_converted <- GSE.156037_converted %>%
  select(Transcript.stable.ID, starts_with("S")) %>%  # select() selects the desired columns in the order you give it, ends_with() is also helpful
  rename("transcript_stable_ID" = Transcript.stable.ID)  # rename the ID column so they are all the same
head(GSE.156037_converted)

# check retention
sprintf("original number of genes: %i", nrow(GSE.156037))
sprintf("converted number of genes: %i", nrow(GSE.156037_converted))
sprintf("proportion conserved: %f", nrow(GSE.156037_converted) / nrow(GSE.156037))
```


Select only counts columns and merge samples with different numbers of genes. Merge with gene ID table to exclude gene identifiers not in table. *This step is important for post-WGCNA functional annotation and analysis.
```{r GSE41205 cleanup}
# check
head(GSE.41205_N2)
head(GSE.41205_KU25)

# select only counts columns, rename gene ID column 
GSE.41205_N2 <- GSE.41205_N2 %>%
  select("GeneID", starts_with("Uniq_reads")) %>%  # the raw counts columns begin with Uniq_reads
  rename("transcript_stable_ID" = GeneID)
GSE.41205_KU25 <- GSE.41205_KU25 %>%
  select("GeneID", starts_with("Uniq_reads")) %>%
  rename("transcript_stable_ID" = GeneID)

# need to use merge function because different number of rows
GSE.41205_merged <- merge(GSE.41205_N2, GSE.41205_KU25, by="transcript_stable_ID")

# exclude genes that are not in the ID table 
GSE.41205_merged
GSE.41205_converted <- merge(GSE.41205_merged, ID_table, by.x = "transcript_stable_ID", 
                             by.y = "Transcript.stable.ID", all.x = FALSE, all.y = FALSE)
head(GSE.41205_converted)  # check which columns you want to keep, i.e. only transcript stable ID and raw counts columns

GSE.41205_converted <- GSE.41205_converted %>%
  select(transcript_stable_ID, starts_with("Uniq_reads"))
head(GSE.41205_converted)

# check retention
sprintf("original number of genes: %i", nrow(GSE.41205_merged))
sprintf("converted number of genes: %i", nrow(GSE.41205_converted))
sprintf("proportion conserved: %f", nrow(GSE.41205_converted) / nrow(GSE.41205_merged))
```


Select only counts columns and merge samples with the same number of genes. Merge with gene ID table to exclude gene identifiers not in table. *This step is important for post-WGCNA functional annotation and analysis.
```{r GSE63528 cleanup}
# check
head(GSE.63528_1)

# select only the counts columns
GSE.63528_1 <- GSE.63528_1 %>% select("Count")
GSE.63528_2 <- GSE.63528_2 %>% select("Count")
GSE.63528_3 <- GSE.63528_3 %>% select("Count")
GSE.63528_4 <- GSE.63528_4 %>% select("Count")
GSE.63528_5 <- GSE.63528_5 %>% select("Count")

# group all samples into a list
GSE.63528_list <- c(GSE.63528_1, GSE.63528_2, GSE.63528_3, GSE.63528_4, GSE.63528_5)

# bind columns and convert tibble to df, re-add ID info
GSE.63528_merged <- as.data.frame(bind_cols(GSE.63528_list))
rownames(GSE.63528_merged) <- rownames(GSE.63528_1)
GSE.63528_merged <- rownames_to_column(GSE.63528_merged, var = "transcript_stable_ID")
GSE.63528_merged


# exclude genes that are not in the ID table 
GSE.63528_merged
GSE.63528_converted <- merge(GSE.63528_merged, ID_table, by.x = "transcript_stable_ID", 
                             by.y = "Transcript.stable.ID", all.x = FALSE, all.y = FALSE)
head(GSE.63528_converted)  # check which columns you want to keep, i.e. only transcript stable ID and raw counts columns

GSE.63528_converted <- GSE.63528_converted %>%
  select(transcript_stable_ID, starts_with("Count"))
head(GSE.63528_converted)

# check retention
sprintf("original number of genes: %i", nrow(GSE.63528_merged))
sprintf("converted number of genes: %i", nrow(GSE.63528_converted))
sprintf("proportion conserved: %f", nrow(GSE.63528_converted) / nrow(GSE.63528_merged))
```

Convert Gene Names to Transcript Stable IDs and select only counts columns
```{r GSE46257 cleanup}
GSE.46257  # check the name of the ID column
                          # change by.x to match gene ID column, ***change by.y to Gene.name 
GSE.46257_converted <- merge(GSE.46257, ID_table, by.x = "gene", by.y = "Gene.name", all.x = FALSE, all.y = FALSE)
GSE.46257_converted  # check which columns you want to keep, i.e. only transcript stable ID and raw counts columns

GSE.46257_converted <- GSE.46257_converted %>%
  select(Transcript.stable.ID, ends_with("counts")) %>%  # select() selects the desired columns in the order you give it
  rename("transcript_stable_ID" = Transcript.stable.ID)  # rename the ID column so they are all the same

# check retention
sprintf("original number of genes: %i", nrow(GSE.46257))
sprintf("converted number of genes: %i", nrow(GSE.46257_converted))
sprintf("proportion conserved: %f", nrow(GSE.46257_converted) / nrow(GSE.46257))
```


### Final Check
Final check of the datasets before you write them out. They should have transcript stable IDs with the same column name and only raw counts for each sample.
```{r check before write out}
GSE.156037_converted
GSE.41205_converted
GSE.63528_converted
GSE.46257_converted
```

### Write out
If they all look good, set the directory you want to write to and write them out as csv files.
```{r write out}
# set directory for where files should be written to
setwd("~/bioinformatics/group-projects-invivo-fall-project/dataset_cleanup_tutorial/")

                                # what the file will be named
write_csv(GSE.156037_converted, "GSE156037_cleaned_transcrID.csv")
write_csv(GSE.41205_converted, "GSE41205_cleaned_transcrID.csv")
write_csv(GSE.63528_converted, "GSE63528_cleaned_transcrID.csv")
write_csv(GSE.46257_converted, "GSE46257_cleaned_transcrID.csv")
```

### Conclusion
Once the cleaned datasets have been created, they can be mixed and matched and further processed for WGCNA using the pre-processing_for_WGCNA.Rmd script.


### Additional hypotheticals
What if there are more than 2 samples in multiple files and they all have different gene counts?
- You would need to do successive merges

