---
title: "Exploring Wormcat"
output:
  html_document: default
---

### Notes on running Wormcat package
- The ```plotflow``` package must be loaded to generate bubble plots

- With no path, the annotation file name must match one of the files in https://github.com/dphiggs01/Wormcat/tree/master/inst/extdata

- The input_type must exactly match "Wormbase.ID" or "Sequence.ID"

- The input gene set enrichment IDs must have a column header matching either "Wormbase.ID" or "Sequence.ID" 
  to merge the dataframes

*\* Wormcat does not like tidyverse, restart R/Rstudio after loading tidyverse

```{r libraries, echo=FALSE, message=FALSE, results='hide'}
library(wormcat)

# plotflow needed to generate bubbleplots
library(plotflow)
```

### Example using Formatted_WormCat_input.txt

```{r test_wormcat, results='hold', echo=TRUE, include=FALSE, eval=FALSE}
input_file <- "~/bioinformatics/group-projects-invivo-fall-project/WormCat_files/Formatted_WormCat_input.txt"

worm_cat_fun(input_file, annotation_file ="whole_genome_jul-03-2019.csv", input_type = "Wormbase.ID")
```


### Output

- A zipped and an unzipped directory, both with the same files

- A csv file with the input genes and annotations
  - rgs_and_categories.csv

- Csv files with Fisher's test information for each category
  - rgs_fisher_cat1.csv
  - rgs_fisher_cat2.csv
  - rgs_fisher_cat3.csv

- Csv files with information for the bubble plots
  - rgs_fisher_cat1_apv.csv
  - rgs_fisher_cat2_apv.csv
  - rgs_fisher_cat3_apv.csv

- Svg files with the bubble plots
  - rgs_fisher_cat1_apv.svg

*\* Sunburst plot is not included


### Playing with the treatment data

```{r install packages, echo=FALSE, message=FALSE, include=FALSE, results='hide'}
#Install 

if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if(!require(biomaRt)) BiocManager::install("biomaRt")
if(!require(edgeR)) BiocManager::install("edgeR")
if(!require(EnhancedVolcano)) BiocManager::install('EnhancedVolcano')
if(!require(DESeq2)) BiocManager::install("DESeq2")
if(!require(GO.db)) BiocManager::install("GO.db")
if(!require(org.Ce.eg.db)) BiocManager::install("org.Ce.eg.db")
if(!require(VennDiagram)) install.packages("VennDiagram", repos = "http://cran.us.r-project.org")
if(!require(RColorBrewer)) install.packages("RColorBrewer", repos = "http://cran.us.r-project.org")
if(!require(formattable)) install.packages("formattable", repos = "http://cran.us.r-project.org")

library(edgeR)
library(biomaRt)
library(org.Ce.eg.db) # bioconductor package genome wide annotation for worm
library(GO.db)
library(EnhancedVolcano)
library(DESeq2)
library(RColorBrewer)
library(VennDiagram)
library(gridExtra)
library(grid)
library(formattable)

setwd('.')
```

```{r read in files, echo=FALSE, include=FALSE, results='hide'}
#reading in counts data, removing last 5 lines that don't contain gene counts
GenewiseCounts <- read.table("~/bioinformatics/group-projects-invivo-fall-project/BGMP_IVB_2020_dataset/merged_counts.tsv",
                             header = TRUE, row.names = 1, sep="\t", check.names = FALSE)
GenewiseCounts <- GenewiseCounts[1:(nrow(GenewiseCounts)-5),]

#setting groups
conditions <- colnames(GenewiseCounts)
group <- vector()
for (cond in conditions) {
  if (grepl("_A", cond)) {
    group <- append(group, "group_A")
  }
  else if (grepl("_B", cond)) {
    group <- append(group, "group_B")
  }
  else if (grepl("_C", cond)) {
    group <- append(group, "group_C")
  }
  else if (grepl("_D", cond)) {
    group <- append(group, "group_D")
  }
}

#determining number of pairwise comparisons
if ("group_D" %in% group) {
  group_num <- 4
} else if ("group_C" %in% group) {
  group_num <- 3
} else if ("group_B" %in% group) {
  group_num <- 2
}
  
print("The number of condition groups submitted:")
print(group_num)
```

```{r remove rRNA, echo=FALSE, results='hide'}
#removing rRNA genes from count matrix

#known rRNA genes listed in the Ensembl database
rrna <- c("WBGene00004512",
"WBGene00004513",
"WBGene00004567",
"WBGene00004622",
"WBGene00014454",
"WBGene00014472",
"WBGene00014621",
"WBGene00077465",
"WBGene00077466",
"WBGene00077467",
"WBGene00077468",
"WBGene00077469",
"WBGene00077470",
"WBGene00077471",
"WBGene00077472",
"WBGene00077473",
"WBGene00077474",
"WBGene00077475",
"WBGene00077476",
"WBGene00077477",
"WBGene00189966",
"WBGene00235197")
rrna_df <- data.frame(rrna)
merged_df <- merge(GenewiseCounts, rrna_df, by.x = "row.names", by.y = "rrna")

colnames(merged_df)[1] <- "rRNA Gene Name"

# print("The raw counts for known rRNA genes are outputted in the following table")
# knitr::kable(merged_df, caption = "Raw rRNA Counts")

GenewiseCounts <- GenewiseCounts[!(rownames(GenewiseCounts) %in% rrna),]
```

```{r DGE, include=FALSE, echo=FALSE, results='hide'}
#creating DGE object, getting dispersion, fitting generalized linear model, and identifying gene names for each ensembl id

#setting dif gene expression list and normalizing
dge.er <- DGEList(counts=GenewiseCounts)
dge.er <- calcNormFactors(dge.er)

#setting design variable
design.er <- model.matrix(~0 + group)

#running all dispersion options
dge.er <- estimateGLMCommonDisp(dge.er, design.er)
dge.er <- estimateGLMTrendedDisp(dge.er, design.er)
dge.er <- estimateGLMTagwiseDisp(dge.er, design.er)

#fitting a negative binomial generalized log-linear model
fit.er <- glmFit(dge.er, design.er)

```

```{r venn diagram 3 treats, fig.height = 7, fig.width = 7, message=FALSE, echo=FALSE, results='hide'}
#3 groups
#specifying contrasts, normalizing data, liklihood ratio tests, adding gene names, and producing venn diagram

if (group_num == 3){
  
  #setting all contrasts for DGE (including GO)
  contrasts <- makeContrasts(A_vs_B=groupgroup_A-groupgroup_B, A_vs_C=groupgroup_A-groupgroup_C, B_vs_C=groupgroup_B-groupgroup_C, levels=design.er)
  
  #conducting likelihood ratio tests for each pairwise comparison
  A_vs_B_lrt <- glmLRT(fit.er, contrast=contrasts[,"A_vs_B"])
  A_vs_C_lrt <- glmLRT(fit.er, contrast=contrasts[,"A_vs_C"])
  B_vs_C_lrt <- glmLRT(fit.er, contrast=contrasts[,"B_vs_C"])
  
  #identifying significantly expressed genes based off of trusted edgeR p-values and stats
  A_vs_B_de <- decideTestsDGE(A_vs_B_lrt, adjust.method = "fdr")
  A_vs_C_de <- decideTestsDGE(A_vs_C_lrt, adjust.method = "fdr")
  B_vs_C_de <- decideTestsDGE(B_vs_C_lrt, adjust.method = "fdr")
  
  summary(A_vs_B_de)
  summary(A_vs_C_de)
  summary(B_vs_C_de)
  
  #creating logCPM table containing all samples to be used for heat maps
  logCPM <- cpm(dge.er, prior.count=2, log=TRUE)
  

  #subsetting each pairwise comparison
  logCPM_A_vs_B <- logCPM[,grepl("_A|_B", colnames(logCPM))]
  logCPM_A_vs_C <- logCPM[,grepl("_A|_C", colnames(logCPM))]
  logCPM_B_vs_C <- logCPM[,grepl("_B|_C", colnames(logCPM))]
  
 }
```

```{r A vs B significant, echo=FALSE, results='hide'}
if (group_num == 3) {
  #getting table of values listed as significant in the volcano plot
  FC_sub_A_B <- subset(A_vs_B_lrt$table, abs(logFC) >= 1)
  P_FC_sub_A_B <- subset(FC_sub_A_B, PValue <= 0.00001)
  P_FC_sub_A_B
}
```

```{r A vs C significant, echo=FALSE, results='hide'}
if (group_num == 3) {
  #getting table of values listed as significant in the volcano plot
  FC_sub_A_C <- subset(A_vs_C_lrt$table, abs(logFC) >= 1)
  P_FC_sub_A_C <- subset(FC_sub_A_C, PValue <= 0.00001)
  P_FC_sub_A_C
 }
```

```{r B vs C significant, echo=FALSE, results='hide'}
if (group_num == 3) { 
  #getting table of values listed as significant in the volcano plot
  FC_sub_B_C <- subset(B_vs_C_lrt$table, abs(logFC) >= 1)
  P_FC_sub_B_C <- subset(FC_sub_B_C, PValue <= 0.00001)
  P_FC_sub_B_C
}
```

```{r three WC GO, message=FALSE, echo=FALSE, results='hide'}
#####Setting DGE list for WCGO
WC_dge.er <- DGEList(counts=GenewiseCounts)

#looking at number of genes in dge variable
WC_dge.er <- calcNormFactors(WC_dge.er)

#getting dispersions
#using same design.er as created earlier
WC_dge.er <- estimateGLMCommonDisp(WC_dge.er, design.er)
WC_dge.er <- estimateGLMTrendedDisp(WC_dge.er, design.er)
WC_dge.er <- estimateGLMTagwiseDisp(WC_dge.er, design.er)

#fitting linear model
WC_fit.er <- glmFit(WC_dge.er, design.er)
```

```{r likelihood ratio tests, echo=FALSE, results='hide'}
#conducting likelihood ratio tests for each pairwise comparison
WC_A_vs_B_lrt <- glmLRT(WC_fit.er, contrast=contrasts[,"A_vs_B"])
WC_A_vs_C_lrt <- glmLRT(WC_fit.er, contrast=contrasts[,"A_vs_C"])
WC_B_vs_C_lrt <- glmLRT(WC_fit.er, contrast=contrasts[,"B_vs_C"])

WC_A_vs_B_lrt
WC_A_vs_C_lrt
WC_B_vs_C_lrt

# subset differentially expressed genes
WC_A_vs_B_FC <- subset(WC_A_vs_B_lrt$table, abs(logFC) >= 1)
WC_A_vs_C_FC <- subset(WC_A_vs_C_lrt$table, abs(logFC) >= 1)
WC_B_vs_C_FC <- subset(WC_B_vs_C_lrt$table, abs(logFC) >= 1)

# subset via pvalue
WC_A_vs_B_FC_P <- subset(WC_A_vs_B_FC, PValue <= 0.05)
WC_A_vs_C_FC_P <- subset(WC_A_vs_C_FC, PValue <= 0.05)
WC_B_vs_C_FC_P <- subset(WC_B_vs_C_FC, PValue <= 0.05)

# need to write out the rownames with "Wormbase.ID" as a header
# convert edgeR objects into data frames
WC_A_vs_B_df <- data.frame(WC_A_vs_B_FC_P)
WC_A_vs_C_df <- data.frame(WC_A_vs_C_FC_P)
WC_B_vs_C_df <- data.frame(WC_B_vs_C_FC_P)

# add "Wormbase.ID" to the beginning of the vector
WC_A_vs_B_IDs <- c("Wormbase.ID", row.names(WC_A_vs_B_df))
WC_A_vs_C_IDs <- c("Wormbase.ID", row.names(WC_A_vs_C_df))
WC_B_vs_C_IDs <- c("Wormbase.ID", row.names(WC_B_vs_C_df))

# number of DEGs with p-value <= 0.05, but NOT FDR corrected
length(WC_A_vs_B_IDs)
length(WC_A_vs_C_IDs)
length(WC_B_vs_C_IDs)

# write out gene lists to be read into wormcat
writeLines(WC_A_vs_B_IDs, con = "WormCat_A_vs_B.csv")
writeLines(WC_A_vs_C_IDs, con = "WormCat_A_vs_C.csv")
writeLines(WC_B_vs_C_IDs, con = "WormCat_B_vs_C.csv")
```


```{r wormcat, echo=FALSE, include=FALSE, results='hide', eval=FALSE}
# execute WormCat
worm_cat_fun("./WormCat_A_vs_B.csv", annotation_file ="whole_genome_jul-03-2019.csv",
             input_type = "Wormbase.ID")

worm_cat_fun("./WormCat_A_vs_C.csv", annotation_file ="whole_genome_jul-03-2019.csv",
             input_type = "Wormbase.ID")

worm_cat_fun("./WormCat_B_vs_C.csv", annotation_file ="whole_genome_jul-03-2019.csv",
             input_type = "Wormbase.ID")
```




