---
title: "Pre-processing for WGCNA"
output: html_document
---

This script is for merging cleaned datasets (see Dataset_Cleanup_Tutorial.Rmd for cleanup information), removing rRNA genes, normalization, and creating the design matrix.

```{r set directory}
knitr::opts_knit$set(root.dir = '~/bioinformatics/group-projects-invivo-fall-project/WGCNA')
```

```{r}
library(tidyverse)
```

```{bash}
pwd
```

Read in cleaned datasets. Cleaned datasets should all have gene IDs of the same type in the first column and subsequent columns should only contain raw gene counts for each sample.
```{r read in datasets}
# datasets
GSE.55988_transcr <- read.csv("../Cleaned_Datasets_transcript_stableID/GSE55988_cleaned_transcrID.csv")
GSE.49662_transcr <- read.csv("../Cleaned_Datasets_transcript_stableID/GSE49662_cleaned_transcrID.csv")
GSE.46051_transcr <- read.csv("../Cleaned_Datasets_transcript_stableID/GSE46051_cleaned_transcrID.csv")
GSE.54853_transcr <- read.csv("../Cleaned_Datasets_transcript_stableID/GSE54853_cleaned_transcrID.csv")
invivo_transcr <- read.csv("../Cleaned_Datasets_transcript_stableID/invivo_cleaned_transcrID.csv")

# check that they all look good
head(GSE.55988_transcr)
head(GSE.49662_transcr)
head(GSE.46051_transcr)
head(GSE.54853_transcr)
head(invivo_transcr)
sprintf("number of genes: %i, number of samples: %i", nrow(GSE.54853_transcr), (ncol(GSE.54853_transcr)-1))
sprintf("number of genes: %i, number of samples: %i", nrow(GSE.49662_transcr), (ncol(GSE.49662_transcr)-1))
sprintf("number of genes: %i, number of samples: %i", nrow(GSE.46051_transcr), (ncol(GSE.46051_transcr)-1))
sprintf("number of genes: %i, number of samples: %i", nrow(GSE.54853_transcr), (ncol(GSE.54853_transcr)-1))
sprintf("number of genes: %i, number of samples: %i", nrow(invivo_transcr), (ncol(invivo_transcr)-1))
```

This chunk merges the datasets. Put the first 2 datasets in the first function, then incorporate next dataset with the previously merged dataset.
```{r merge}
# merge all files into one dataframe for WGCNA; the order for merging is determined based on the order of the samples in the design matrix
merged1 <- merge(GSE.55988_transcr,GSE.49662_transcr,by='transcript_stable_ID',all=TRUE)
merged2 <- merge(merged1,GSE.46051_transcr,by='transcript_stable_ID',all=TRUE)
merged3 <- merge(merged2,GSE.54853_transcr,by='transcript_stable_ID',all=TRUE)
merged_final <- merge(merged3,invivo_transcr,by='transcript_stable_ID',all=TRUE)
merged_final[is.na(merged_final)] <- 0
row.names(merged_final) <- merged_final[,1]
merged_final[,1] <- NULL

# check merge
merged_final
sprintf("number of genes: %i, number of samples: %i", nrow(merged_final), (ncol(merged_final)-1))
```

Remove known rRNA genes
```{r remove rRNA genes}
#known rRNA genes listed in the Ensembl database, release 102
rrna <- c("F31C3.7", "rrn-1.1",
          "F31C3.8", "rrn-1.2",
          "F31C3.11", "rrn-2.1",
          "F31C3.9", "rrn-3.1",
          "MTCE.7", "MTCE.7",
          "MTCE.33", "MTCE.33",
          "ZK218.12", "rrn-4.1",
          "ZK218.16", "rrn-4.2",
          "ZK218.17", "rrn-4.3",
          "ZK218.18", "rrn-4.4",
          "ZK218.19", "rrn-4.5",
          "ZK218.20", "rrn-4.6",
          "Y102A5D.5", "rrn-4.7",
          "Y102A5D.6", "rrn-4.8",
          "Y102A5D.7", "rrn-4.9",
          "Y102A5D.8", "rrn-4.10",
          "Y102A5D.9", "rrn-4.11",
          "Y102A5D.10", "rrn-4.12",
          "Y102A5D.11", "rrn-4.13",
          "Y102A5D.12", "rrn-4.14",
          "T27C5.18", "rrn-4.15",
          "T09B4.23", "rrn-4.16")

rrna_df <- data.frame(rrna)
rrna_counts <- merge(merged_final, rrna_df, by.x = "row.names", by.y = "rrna")

colnames(rrna_counts)[1] <- "rRNA Gene Name"

print("The raw counts for known rRNA genes are outputted in the following table")
knitr::kable(rrna_counts, caption = "Raw rRNA Counts")

# remove rRNA counts
merged_final_no_rRNA <- merged_final[!(rownames(merged_final) %in% rrna),]
merged_final_no_rRNA

sprintf("number of genes conserved after rRNA removal: %i", nrow(merged_final_no_rRNA))
```

Comment back in for writing out raw counts
<!-- Write out raw counts -->
<!-- ```{r write out merged raw counts} -->
<!-- setwd("~/bioinformatics/group-projects-invivo-fall-project/WGCNA/") -->

<!-- writeout <- rownames_to_column(merged_final_no_rRNA, var = "transcript_stable_ID") -->
<!-- write.csv(writeout, "raw_wormspellGEO_transcript_stable_ID.csv") -->
<!-- ``` -->

## Normalization

```{r TPM normalization}
# get length data
length_data <- read.table("../BGMP_IVB_2020_dataset/GC_lengths.tsv",
                          header = TRUE, sep="\t", check.names = FALSE)
length_data <- length_data[!(rownames(length_data) %in% rrna),]   # remove rRNA genes from length data
length_data$Length <- as.numeric(length_data$Length)
length_data$Length_kb <- length_data$Length / 1000   # store length in kb

# convert length data IDs
ID_table <- read.csv("../Cleaned_Datasets_transcript_stableID/celegans_cleaned_mart_ensembl.102.tsv", sep = "\t")
length_data <- rownames_to_column(length_data, var = "Gene.stable.ID")
lengthdata_converted <- merge(length_data, ID_table, by="Gene.stable.ID", all.x = FALSE, all.y = FALSE)
lengthdata_converted <- lengthdata_converted %>%
  select(Transcript.stable.ID, Length_kb) %>%
  rename("transcript_stable_ID" = Transcript.stable.ID)
lengthdata_converted


merged_with_lengths <- merge(merged_final_no_rRNA, lengthdata_converted, by.x=0, by.y="transcript_stable_ID", sort = TRUE)
counts_per_kb <- merged_with_lengths[,2:(ncol(merged_with_lengths)-1)]/merged_with_lengths[,(ncol(merged_with_lengths))]
rownames(counts_per_kb) <- merged_with_lengths$Row.names
scale_factors <- (colSums(counts_per_kb) / 10^6)
tpm_normalized_df <- counts_per_kb / scale_factors
tpm_normalized_df

# log transformation
tmp_withLogtrans <- log2((tpm_normalized_df+1))
tmp_withLogtrans
```

Write out TPM normalized counts
```{r write out merged raw counts}
setwd("~/bioinformatics/group-projects-invivo-fall-project/WGCNA/")

writeout <- rownames_to_column(merged_final_no_rRNA, var = "transcript_stable_ID")
write.csv(writeout, "TPM_wormspellGEO_transcript_stable_ID.csv")
```


## Design Matrix

Build the experimental design matrix from a metadata file. The easiest way to create this metadata file
is to use Excel or a similar application and save as a CSV. The metadata file should contain the 
following columns: 1) Sample -> contains the descriptive name for each sample in the dataset, 2) 
Condition -> contains the experimental condition for which the sample was subjected (i.e., Heat Stress, 
DR, etc.), and 3) Label -> contains the original label provided in the counts dataset pulled from 
GEO (or any other database). **Note:** It is important to ensure that the order of the samples (columns) 
in the merged expression dataset is exactly the same as the order of the samples in the 'Sample' column 
of the experiment metadata file. Changing the order will compromise the results from WGCNA.
```{r}
# import the experimental metadata file
experimental_info <- read.csv('experiment_metadata.csv',header = TRUE)
# build the design matrix from the sample and condition columns
design <- model.matrix(Sample ~ 0 + Condition,data = experimental_info)
# rename the columns and assign the row names
colnames(design) <- sub('^Condition','',colnames(design))
rownames(design) <- experimental_info$Sample
# verify the design
head(design)
# write out for use in WGCNA
write.csv(design,'design_matrix.csv')
```



