---
title: "WGCNA"
output: html_document
---

```{r setup, include=FALSE, echo=FALSE}
# set global parameters for the file
require("knitr")
opts_knit$set(root.dir = ".")
fig.dim <- 6
knitr::opts_chunk$set(echo = FALSE,
                      fig.width=2*fig.dim,
                      fig.height=fig.dim,
                      fig.align='center')
```

Load required packages for analysis.
```{r load packages, message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if(!require(flashClust)) BiocManager::install("flashClust")
if(!require(WGCNA)) BiocManager::install("WGCNA")
if(!require(edgeR)) BiocManager::install("edgeR")
# Load WGCNA and flashClust libraries every time you open R
library(WGCNA)
library(flashClust)
library(edgeR)
# this is an important parameter for WGCNA analysis
options(stringsAsFactors = FALSE)
```

Load raw expression data containing genes as rows and samples as columns.
```{r get expression data}
# set the gene ID label being used in this experiment (i.e., external_gene_name, wormbase_gene, transcript_stable_ID, etc.)
gene_ids <- 'transcript_stable_ID'
# read in data and set gene names as rownames
datExpr <- read.csv("TPM_wormspellGEO_transcript_stable_ID.csv", row.names = gene_ids)
datExpr$X <- NULL
# "head" the file to preview it
head(datExpr)
```

Modify the expression data so it is compatible with WGCNA processing (i.e., transpose dataframe 
so that samples are now rows and genes are columns). Remove genes and samples that have excessive 
missing entries or zero variance. Parameters for 'goodSamplesGenes' function can be modified to 
meet the needs of the data.
```{r remove genes with missing values or no variance, results='hide'}
datExpr <- as.data.frame(t(datExpr))
rownames(datExpr) <- sub('^X','',rownames(datExpr))
dim(datExpr)

# Run this to check if there are gene outliers
gsg <- goodSamplesGenes(datExpr, verbose = 3)
gsg$allOK 
#If the last statement returns TRUE, all genes have passed the cuts. If not, we remove the offending genes and samples from the data with the following:
if (!gsg$allOK)
  {if (sum(!gsg$goodGenes)>0)
		printFlush(paste("Removing genes:", paste(names(datExpr)[!gsg$goodGenes], collapse = ", ")));
		if (sum(!gsg$goodSamples)>0)
			printFlush(paste("Removing samples:", paste(rownames(datExpr)[!gsg$goodSamples], collapse = ", ")))
		datExpr <- datExpr[gsg$goodSamples, gsg$goodGenes]
		}
```

This section provides a sample examination to ensure that all samples are satisfactory for analysis, i.e., 
there are no outlier samples present in the dataset.
```{r check for outlier samples}
sampleTree <- hclust(dist(datExpr), method = "average")
# set the parameters for the plot
sizeGrWindow(12,9)
par(cex = 0.6)
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Sample clustering", sub = "", xlab = "", cex.lab = 1.5,
cex.axis = 1.5, cex.main = 2)
```

Import the design matrix containing the condition information, mapping samples to specific conditions 
or treatments used in this dataset. The sample names from the expression data frame 'datExpr' is mapped 
onto the new data frame 'datTraits' so even if the names are different in the file being imported, this 
code chunk will make them match.
```{r get trait data}
# Create an object called "datTraits" that contains your trait data
datTraits <- read.csv("design_matrix.csv")
# modify the expression data frame so that it matches the trait sample names
rownames(datTraits) <- datTraits$X
datTraits$X <- NULL
# make the sample names in both the expression dataset and traits dataset match
rownames(datExpr) <- rownames(datTraits)
# check out the traits data
head(datTraits)
# ensure rownames for both the expression dataset and condition (design) dataset are equal
table(rownames(datTraits) == rownames(datExpr))
```

Save the processed and cleaned datasets for use in downstream analysis.
```{r save data}
save(datExpr, datTraits, file = "expression_trait_data.RData")
```

This code chunk chooses a soft thresholding power for which the pairwise gene correlations are raised 
to amplify the differences between strong and weak correlations present in the data. This code takes 
a few minutes to run. A modified version of the biweight midcorrelation function is used ('bicor'), 
which has been found to work well with gene expression data. 
```{r soft thresholding calculations}
powers <- c(c(1:10), seq(from = 10, to = 20, by = 1)) # choosing a set of soft-thresholding powers
sft <- pickSoftThreshold(datExpr, powerVector = powers, verbose = 5, networkType = "signed") # call network topology analysis function
```

This plot shows the scale free topology model fit (R^2) for each of the thresholds chosen above. The idea 
is to choose the smallest threshold that reaches the 0.9 correlation cutoff value (red line in plot). The 
model fit (R^2) represents how well the connections within the graph approximate a scale free topology (i.e., 
few nodes with many connections and many nodes with few connections).
```{r scale independence plot}
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], xlab = "Soft Threshold (power)", ylab = "Scale Free Topology Model Fit, signed R^2",
     type = "n", main = paste("Scale independence"))
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], labels = powers, col = "red")
abline(h = 0.90, col = "red")
```

This plot shows the mean connectivity for nodes in the network. Although the previous graph provides the 
more important metric for selecting a soft thresholding power, a higher mean connectivity is desired for 
network analysis. As shown by the two graphs, there is a trade-off, so the best threshold is a balance 
between the two.
```{r mean connectivity plot}
plot(sft$fitIndices[,1], sft$fitIndices[,5], xlab = "Soft Threshold (power)", ylab = "Mean Connectivity", type = "n", main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels = powers, col = "red")
```

Based on our sample dataset (Wormbase Spell), a soft threshold power value of 8 is appropriate for network 
analysis. **Note:** For future datasets, change the 'softPower' variable to reflect the chosen power value 
based on the above graphs. 
```{r get adjacency matrix}
softPower <- 8
adjacency <- adjacency(datExpr, power = softPower, type = "signed") #specify network type, we care about the direction of change so the 'type' is set to 'signed'; if direction of correlation is not important, set to 'unsigned'
```

Construct the network by computing the topological overlap matrix and the corresponding dissimilarity values. 
This calculation takes the gene correlations and provides topology information to build the network.
```{r get TOM matrix}
TOM <- TOMsimilarity(adjacency, TOMType = "signed") # specify network type; make sure its the same specification as provided in the 'adjacency' function call above
dissTOM <- 1-TOM
rm(adjacency) # free up some memory
```

This code chunk generates the modules from the network. Only parameters set globally need to be changed, for 
example, 'minModuleSize' can be changed to meet the needs of the analysis. Other parameters within the functions 
can stay the same for now. Additionally, the second plot shows a dendrogram of module dissimilarity. This can 
be used to set thresholds for merging module with similar expression profiles. 
```{r generate modules}
# Generate a clustered gene tree
geneTree <- flashClust(as.dist(dissTOM), method = "average")
# plot the dendrogram
plot(geneTree, xlab = "", sub = "", main = "Gene Clustering on TOM-based dissimilarity", labels = FALSE, hang = 0.04)

# This sets the minimum number of genes to cluster into a module
minModuleSize <- 100
# this function is used to build the modules
dynamicMods <- cutreeDynamic(dendro = geneTree, distM = dissTOM, deepSplit = 2, pamRespectsDendro = FALSE, minClusterSize = minModuleSize)

# get attributes of the modules, such as module colors, eigengenes, and module dissimilarity measures used for plotting dendrogram
dynamicColors <- labels2colors(dynamicMods)
MEList <- moduleEigengenes(datExpr, colors = dynamicColors,softPower = softPower)
MEs <- MEList$eigengenes
MEDiss <- 1-cor(MEs)
METree <- flashClust(as.dist(MEDiss), method = "average")

# plots tree showing how the eigengenes cluster together
plot(METree, main = "Clustering of module eigengenes", xlab = "", sub = "")
```

Choose a dissimilarity threshold for merging similar modules (0.25 is suggested so will be the default here, 
but can be changed to any value based on above eigengene dendrogram), and plot the corresponding dendrogram 
with module assignment below. 
```{r merge similar modules}
# set a threhold for merging modules. if merging is not desired, set the value to 0.0
MEDissThres <- 0.25
merge <- mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
mergedColors <- merge$colors
mergedMEs <- merge$newMEs

# plot dendrogram with module colors below it
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors), c("Dynamic Tree Cut", "Merged dynamic"), dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05)

# update the objects defined pre-merging with the merged information
moduleColors <- mergedColors
colorOrder <- c("grey", standardColors(50))
moduleLabels <- match(moduleColors, colorOrder)-1
MEs <- mergedMEs
MEDiss <- 1-cor(MEs)
```

Correlate the modules to external traits in the dataset.
```{r correlate modules to traits}
# Define number of genes and samples
nGenes <- ncol(datExpr)
nSamples <- nrow(datExpr)

# Recalculate MEs with color labels
MEs0 <- moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs <- orderMEs(MEs0)
moduleTraitCor <- cor(MEs, datTraits, use = "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nSamples)

# Print correlation heatmap between modules and traits
textMatrix <- paste(signif(moduleTraitCor, 2), "\n(", 
                    signif(moduleTraitPvalue, 1), ")", sep = "")
dim(textMatrix) <- dim(moduleTraitCor)
# par(mar = c(10, 10, 4, 1))
# tiff('module_trait_relationship.tif',width=6,height=5,units='in',res = 300)
# display the corelation values with a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor, 
               xLabels = names(datTraits), 
               yLabels = names(MEs), 
               ySymbols = names(MEs), 
               colorLabels = FALSE, 
               colors = blueWhiteRed(50), 
               textMatrix = textMatrix,
               cex.lab.x = 0.6,
               cex.lab.y = 0.6,
               xLabelsAngle = 90,
               setStdMargins = FALSE, 
               cex.lab = 0.9,
               cex.text = 0.3, 
               zlim = c(-1,1), 
               main = "Module-trait relationships")
# dev.off()
```

Pull out the interesting modules and the traits they correlate to for downstream analysis.  
```{r get interesting modules}
# set a correlation threshold for selecting interesting modules and the traits they highly correlate to
corThreshold <- 0.75
# get the module colors and trait names
indexes <- which(abs(moduleTraitCor) >= corThreshold, arr.ind = TRUE)
modules <- rownames(indexes)
traits <- colnames(moduleTraitCor[,indexes[,2],drop=FALSE])
modules
traits
```

With the current parameters and dataset, it can be seen from the heatmap above that the 'darkgreen' 
module correlates highly to the 'Group_C' condition (InVivo provided transcriptomic profile). Save 
the module and trait information for identifying interesting genes and pathways.
```{r save module information}
# save the module information for downstream analysis
save(MEs, moduleColors, modules, traits, file = "module_information.RData")
```